### PR TITLE
Add period syntax to PICMI

### PR DESCRIPTION
**Pull request by chillenzer**
This PR adds an interface to the period syntax for plugins to PICMI. In the process of doing so, I've found a bug that is fixed separately in #5309. This PR also restructures our PICMI implementation by adding a separate folder for `diagnostics` as discussed in #5316.

After a number of discussions, I decided to let the semantics of slicing/indexing slightly deviate from Python's interpretation by making it INCLUSIVE on the upper end as well. That is in line with what warpx is doing and feels natural for time intervals where you would expect to cover the whole range from start to end.

Hints for review:
- The main feature is implemented and documented in `lib/python/picongpu/picmi/diagnostics/timestepspec.py`. It forwards to the rather minimalistic PyPIConGPU pendant.
- The interface has an extensive test suite that should clarify the semantics of slicing, etc. Feel free to (a) consult it in case of uncertainties and (b) suggest further edge cases if you find some.
- Another big source of changes is the re-organisation of the directory structure. That's not very important.
- Finally the feature unit conversions are lazily evaluated at `.get_as_pypicongpu`, so that function needs the time step size and number of steps, so there are some interface changes in that regard propagating through plugins into `simulation.py`.
- I've seen that sometimes the auto-formatter kicked in. If that's too annoying I can try to remove those changes.

### CODE REVIEW

File: lib/python/picongpu/pypicongpu/output/timestepspec.py [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+
+
+def _serialize(spec):
+    if isinstance(spec, slice):
+        return {
+            "start": spec.start if spec.start is not None else 0,
+            "stop": spec.stop if spec.stop is not None else -1,
+            "step": spec.step if spec.step is not None else 1,
+        }
+    raise ValueError(f"Unknown serialization for {spec=} as a time step specifier (--period argument).")
```
Comment (BrianMarre) [inline lib/python/picongpu/pypicongpu/output/timestepspec.py:21]: Wouldn't passing a list[int] cause a ValueError here? Or is  list[int] a slice object for some reason? If yes please add a comment explaining that.

But according to the implementation of TimeStepSpec that should be a valid input value

File: share/picongpu/pypicongpu/examples/laser_wakefield/main.py [hunk 2]
```diff
@@ -107,14 +120,19 @@
...
 
 diagnostics_list = []
 diagnostics_list.append(
-    picmi.PhaseSpace(
+    picmi.diagnostics.PhaseSpace(
         species=electrons,
-        period=100,
+        # Resulting values for period:
+        # 17, 34, 50, 51, 57, 64, 68, 71, 85, 100, 102, ...
+        period=picmi.diagnostics.TimeStepSpec[::100, 50:72:7, 17],
```
Comment (BrianMarre) [inline share/picongpu/pypicongpu/examples/laser_wakefield/main.py]: the syntax for the entry 17 is in my opinion misleading, I would assume that that indicated that we will write an output at the step 17, not every 17 steps as the implementation suggests.

Also the step 0 is missing from the list.

Comment (ikbuibui) [reply share/picongpu/pypicongpu/examples/laser_wakefield/main.py]: 17 really means every 17 steps and not just only timestep number 17

File: lib/python/picongpu/picmi/diagnostics/timestepspec.py [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+# The following might look slightly convoluted but is preferred over the more obvious use of `__class_getitem__`.
+# This is because the latter is supposed to return a GenericAlias which is not what we want.
+# That would be like list[int].
+# We are more like an `Enum` where indexing into the class means something different.
+# See [here](https://docs.python.org/3/reference/datamodel.html#object.__getitem__) and
+# [here](https://docs.python.org/3/reference/datamodel.html#classgetitem-versus-getitem).
+
+
+class _TimeStepSpecMeta(type):
+    # Provide this to have a nice syntax picmi.diagnostics.TimeStepSpec[10:200:5, [3,7,11,17]]
```
Comment (BrianMarre) [inline lib/python/picongpu/picmi/diagnostics/timestepspec.py]: This comment is misleading since list[int] is not supported, rather only slices or integers

File: lib/python/picongpu/pypicongpu/output/timestepspec.py [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+
+
+class TimeStepSpec(RenderedObject):
+    specs = build_typesafe_property(list[slice | list[int]])
```
Comment (BrianMarre) [inline lib/python/picongpu/pypicongpu/output/timestepspec.py]: allowing a list of integers looks to be remainder of a previous implementation, the current one actually only seems to work with list of slices and I consider that a good thing.

File: share/picongpu/pypicongpu/examples/laser_wakefield/main.py (hunk 2 above)
Comment (chillenzer) [reply share/picongpu/pypicongpu/examples/laser_wakefield/main.py]: @ikbuibui is correct but we agreed to change this behaviour. I'll make this mean just time step 17.

File: lib/python/picongpu/pypicongpu/output/timestepspec.py (hunk 4 above)
Comment (chillenzer) [reply lib/python/picongpu/pypicongpu/output/timestepspec.py]: You are correct. Thanks for catching that.

File: lib/python/picongpu/picmi/diagnostics/timestepspec.py [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+            )
+        self.unit_system = unit_system
+        if unit_system == "seconds":
+            self.specs_in_seconds = self.specs
+            self.specs = tuple()
+        return self
+
+    def __add__(self, other):
+        if not (isinstance(other, TimeStepSpec)):
+            raise TypeError("unsupported operand type(s) for +: TimeStepSpec and {type(other)}")
```
Comment (ikbuibui) [inline lib/python/picongpu/picmi/diagnostics/timestepspec.py]:
```suggestion
            raise TypeError(f"unsupported operand type(s) for +: TimeStepSpec and {type(other)}")
```

File: lib/python/picongpu/picmi/diagnostics/timestepspec.py [hunk 6]
```diff
@@ -0,0 +1,198 @@
...
+        # The following guards against setting units on the result of the addition.
+        # Otherwise one could specify time steps in "steps" unit, add that to
+        # another TimeStepSpec and reset the units.
+        ts.unit_system = "mixed"
+        return ts
+
+    def _transform_to_steps(self, specs_in_seconds, time_step_size):
+        return tuple(
+            slice(
+                int(spec.start / time_step_size if spec.start is not None else 0),
```
Comment (ikbuibui) [inline lib/python/picongpu/picmi/diagnostics/timestepspec.py:165]: There should probably be a check somewhere to ensure ``time_step_size > 0``. I say probably, since a subset of all possible simulations could possibly be time reversal invariant. 
It should definitely not be zero though, and this should be checked.

File: lib/python/picongpu/pypicongpu/output/timestepspec.py (hunk 1 above)
Comment (chillenzer) [reply lib/python/picongpu/pypicongpu/output/timestepspec.py:21]: Resolving on behalf of @BrianMarre

File: share/picongpu/pypicongpu/examples/laser_wakefield/main.py (hunk 2 above)
Comment (chillenzer) [reply share/picongpu/pypicongpu/examples/laser_wakefield/main.py]: Resolving on behalf of @BrianMarre

File: lib/python/picongpu/picmi/diagnostics/timestepspec.py (hunk 3 above)
Comment (chillenzer) [reply lib/python/picongpu/picmi/diagnostics/timestepspec.py]: Resolving on behalf of @BrianMarre

File: lib/python/picongpu/pypicongpu/output/timestepspec.py (hunk 4 above)
Comment (chillenzer) [reply lib/python/picongpu/pypicongpu/output/timestepspec.py]: Resolving on behalf of @BrianMarre

**Review by ikbuibui — DISMISSED** [review]
Small comments, nothing blocking.

File: lib/python/picongpu/picmi/diagnostics/timestepspec.py (hunk 5 above)
Comment (chillenzer) [reply lib/python/picongpu/picmi/diagnostics/timestepspec.py]: Thanks for catching this.

File: lib/python/picongpu/picmi/diagnostics/timestepspec.py (hunk 6 above)
Comment (chillenzer) [reply lib/python/picongpu/picmi/diagnostics/timestepspec.py:165]: I've added a check to `_transform_to_steps` because that is one place where `time_step_size == 0` would actually become a problem. But I think it's actually the responsibility of `picmi.Simulation` to check arguments for consistency. I'll add that to my todo list.

**Review by ikbuibui — APPROVED** [review]
